telegram/message/peer/resolver.go

```diff
@@ -28,13 +38,65 @@ func DefaultResolver(raw *tg.Client) Resolver {
 
 type plainResolver struct {
 	raw *tg.Client
+
+	contactsMux  sync.Mutex
+	contacts     *tg.ContactsContacts
+	contactsHash atomic.Int32
 }
 
-func (p plainResolver) Resolve(ctx context.Context, domain string) (tg.InputPeerClass, error) {
+func (p *plainResolver) Resolve(ctx context.Context, domain string) (tg.InputPeerClass, error) {
 	peer, err := p.raw.ContactsResolveUsername(ctx, domain)
 	if err != nil {
 		return nil, xerrors.Errorf("resolve: %w", err)
 	}
 
 	return EntitiesFromResult(peer).ExtractPeer(peer.Peer)
 }
+
+func (p *plainResolver) ResolvePhone(ctx context.Context, phone string) (tg.InputPeerClass, error) {
+	r, err := p.raw.ContactsGetContacts(ctx, int(p.contactsHash.Load()))
+	if err != nil {
+		return nil, xerrors.Errorf("get contacts: %w", err)
+	}
+
+	p.contactsMux.Lock()
+	defer p.contactsMux.Unlock()
+
+	switch c := r.(type) {
+	case *tg.ContactsContacts:
+		p.contacts = c
+		cts := c.Contacts
+
+		sort.SliceStable(cts, func(i, j int) bool {
+			return cts[i].UserID < cts[j].UserID
+		})
+		h := hasher.Hasher{}
+		for _, contact := range cts {
+			h.Update(uint32(contact.UserID))
+		}
+
+		p.contactsHash.Store(h.Sum())
+	case *tg.ContactsContactsNotModified:
+		if p.contacts == nil {
+			return nil, xerrors.Errorf("got unexpected %T result", r)
+		}
```

Comment:
What will happen if contacts list is empty and we pass zero hash?

Reply:
Yep, just checked that and at least for me Telegram returns contacts response with empty list.